lib/spack/spack/spec.py

```diff
@@ -1384,7 +1384,12 @@ def cshort_spec(self):
     @property
     def prefix(self):
         if self._prefix is None:
-            self.prefix = spack.store.layout.path_for_spec(self)
+            upstream, record = spack.store.db.query_by_spec_hash(
+                self.dag_hash())
+            if upstream:
```

Comment:
I think this check should be on record, not upstream. Whether we found the spec locally or upstream, if we found a record we should use the path from that record, and only compute the path when the spec isn't found.

This will be important for things like if a user changes their install layout. They should still be able to see the packages they installed before the change, but they won't if we recompute the prefix when we could've gotten it from the database.

Reply:
If I understand correctly, you are describing the case that a user installs some packages with Spack, and then alters the `install_path_scheme`.

I agree that what you suggest is necessary to support changing the install layout, but I don't know if it is sufficient - are you more certain about that? I'd prefer to handle that in a different PR.

Additionally, this would likely require some additional care. For one, `if not record.installed` then the path should also be taken from `spack.store.layout`.